src/types.rs

```diff
@@ -764,3 +774,190 @@ impl LateLintPass for AbsurdExtremeComparisons {
         }
     }
 }
+
+/// **What it does:** This lint checks for comparisons where the relation is always either true or false, but where one side has been upcast so that the comparison is necessary. Only integer types are checked.
+///
+/// **Why is this bad?** An expression like `let x : u8 = ...; (x as u32) > 300` will mistakenly imply that it is possible for `x` to be outside the range of `u8`.
+///
+/// **Known problems:** None
+///
+/// **Example:** `let x : u8 = ...; (x as u32) > 300`
+declare_lint! {
+    pub INVALID_UPCAST_COMPARISONS, Warn,
+    "a comparison involving an term's upcasting to be within the range of the other side of the \
+    term is always true or false"
+}
+
+pub struct InvalidUpcastComparisons;
+
+impl LintPass for InvalidUpcastComparisons {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INVALID_UPCAST_COMPARISONS)
+    }
+}
+
+enum FullInt {
+    S(i64),
+    U(u64),
+}
+
+use std;
+use std::cmp::Ordering;
+
+impl FullInt {
+    #[allow(cast_sign_loss)]
+    fn cmp_s_u(s: &i64, u: &u64) -> std::cmp::Ordering {
+        if *s < 0 {
+            Ordering::Less
+        } else if *u > (i64::max_value() as u64) {
+            Ordering::Greater
+        } else {
+            (*s as u64).cmp(u)
+        }
+    }
+}
+
+impl PartialEq for FullInt {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+impl Eq for FullInt {}
+
+impl PartialOrd for FullInt {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(match (self, other) {
+            (&FullInt::S(ref s), &FullInt::S(ref o)) => s.cmp(o),
+            (&FullInt::U(ref s), &FullInt::U(ref o)) => s.cmp(o),
+            (&FullInt::S(ref s), &FullInt::U(ref o)) => Self::cmp_s_u(s, o),
+            (&FullInt::U(ref s), &FullInt::S(ref o)) => Self::cmp_s_u(o, s).reverse(),
```

Comment:
You don’t need all the `ref`s here, `s` and `o` are `Copy` (`i64` and `u64`). Also `cmp_s_u` shouldn’t take its parameter by reference.